Pass and record process_group_name when creating ProcessGroupNCCL

shengbao-zheng opened the pull request:
Summary:
Pass python c10d group_name to c++ ProcessGroupNCCL so that the pg name will be consistent across different layers. 
Also record pg_name in flight recorder entry.

Differential Revision: D55597200




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -433,10 +434,11 @@ struct NCCLTraceBuffer {
   size_t next_ = 0;
   size_t id_ = 0;
-  std::map<size_t, std::vector<uint64_t>> pg_id_to_ranks_;
+  std::map<std::string, std::vector<uint64_t>> pg_name_to_ranks_;
```
technically this would require a major version bump.  i think thats ok to do though.  don't think we have enough customers that care about BC yet.

shengbao-zheng replied:
I will do a version bump in the following PR when deprecating pg_id.

kwen2501 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2481,7 +2481,9 @@ Example::
       .def_readwrite(
           "global_ranks_in_group",
-          &::c10d::ProcessGroupNCCL::Options::global_ranks_in_group);
+          &::c10d::ProcessGroupNCCL::Options::global_ranks_in_group)
+      .def_readwrite(
+          "group_name", &::c10d::ProcessGroupNCCL::Options::group_name);
```
It seems to me this attribute would be better at the ProcessGroup level than ProcessGroupNCCL level.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -430,6 +430,7 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     std::shared_ptr<ProcessGroupNCCL> split_from;
     int64_t split_color{0};
     std::vector<uint64_t> global_ranks_in_group;
+    std::string group_name;
```
Same

kwen2501 reviewed (commented):
Can you consider generalizing the attribute to ProcessGroup level?

shengbao-zheng committed:
Pass and record process_group_name when creating ProcessGroupNCCL (#123117)

Summary:

Pass python c10d group_name to c++ ProcessGroupNCCL so that the pg name will be consistent across different layers.
Also record pg_name in flight recorder entry.

Reviewed By: wconstab

Differential Revision: D55597200
